jenkins-dsl/jenkins_jira_integration.py

```diff
@@ -0,0 +1,352 @@
+        break
+    assert result is not None
+    return result
+
+
+def post_results_to_jira(jira: JIRA, ticket_number: str, branch: str, post_to_jira: bool, ci_results: str) -> None:
+    """
+    :param jira: Connected and authenticated JIRA instance
+    :param ticket_number: as named
+    :param branch: CI branch this job was for
+    :param post_to_jira: Whether we should actively update Jira with the comment or print it to local output
+            for a manual run / debugging
+    :param ci_results: str to post to jira ticket
+    :return:
+    """
+    log('Attempting to connect to jira and get issue: CASSANDRA-' + ticket_number)
+    issue = jira.issue('CASSANDRA-' + ticket_number)
+    log('Checking comments on ' + str(issue))
+    comments = issue.fields.comment.comments
+
+    # There's a little nuance here. We have two motions we could potentially need to go through
+    #   1) We don't have anything on this ticket for this branch by JenkinsBot, so we want to add
+    #   2) We already have an entry on this ticket for this branch by JenkinsBot, so we want to update
+
+    # TODO put the correct account name here
+    my_name = 'JenkinsBot'
+
+    # Walk the comments for any we authored with the branch name of what we're processing. If we find it, update it
+    for comment in comments:
+        if str(comment.author) == my_name:
+            # We know we wrote it; now we need to determine if this comment was for the branch we're currently
+            # processing to determine if we need to update it
+            comment_text = jira.comment('CASSANDRA-' + ticket_number, comment)
+            if 'Branch: ' + branch in comment_text.body:
+                # This is an update
+                if post_to_jira is True:
+                    comment.update(body=ci_results)
+                else:
+                    print('[UPDATE] comment to manually post to Jira for CASSANDRA-' + ticket_number)
+                    print(ci_results)
+                return
+
+    if post_to_jira is True:
+        issue.add_comment('CASSANDRA-' + ticket_number, ci_results)
+    else:
+        print('[ADD] the following comment to Jira for CASSANDRA-' + ticket_number)
+        print(ci_results)
+
+
+def build_local_cache(server: jenkins, branch: str, buildnum: str) -> Dict:
+    """
+    For the input build back as far as we have history, we want to cache the following (k/v):
+        NUMBER:             str number of this build
+        PREVIOUS_NUMBER     str build number of the previous build to walk to when walking the cache
+        failures:           set of test failures seen for this build (note: sets aren't serializable in JSON so we cast to list)
+
+        We use the build number and previous number as strings since they're going to get case to that in the JSON
+        ser/deser anyway, so best not to mix.
+    :return: testnu->
+    """
+    if not os.path.isdir('cache'):
+        os.mkdir('cache')
```

Comment:
should this not be a config or some non-relative place?  If run in a job then the cache will be in the workspace, which gets wiped each run.

Reply:
Maybe. I originally wrote this with the model of it running from a single place repeatedly rather than on jenkins agents that get wiped. If we go the agent route, we also have to wrestle with the fact that each would have their own local history to execute from, meaning it would be quite difficult to build a long-running history for each branch.

Reply:
in oss do we have fixed boxes to run this as a cronjob?  the "cache" aspect kinda depends on that